examples/tag_analysis/timing_channels/src/lib.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree
+
+// This is an example of using tag analysis to verify constant-time security.
+// Use the following flag of MIRAI to enable constant-time verification:
+// MIRAI_FLAGS --constant_time SecretTaintKind
+
+#![feature(const_generics)]
+#![allow(incomplete_features)]
+
+#[macro_use]
+extern crate mirai_annotations;
+
+use mirai_annotations::{TagPropagation, TagPropagationSet};
+
+struct SecretTaintKind<const MASK: TagPropagationSet> {}
+
+const SECRET_TAINT_MASK: TagPropagationSet =
+    tag_propagation_set!(TagPropagation::Equals, TagPropagation::Ne);
+
+type SecretTaint = SecretTaintKind<SECRET_TAINT_MASK>;
+
+pub mod non_constant_time {
+    /// A compare function that is **not** constant-time.
+    pub fn compare(secret: &[i32], public: &[i32], len: usize) -> bool {
+        precondition!(secret.len() == len);
+        precondition!(public.len() == len);
+        let mut i = 0;
+        while i < len {
+            assume!(has_tag!(&secret[i], crate::SecretTaint));
```

Comment:
The assumption here amounts to precondition, so rather use a precondition.

In general, though, we should not expect code to have such preconditions. Rather, the summary of this function will contain a conditional expression where the condition, after refinement, will be tainted. That is the place to give the diagnostic.

Reply:
Removed the assumption, but added a precondition about the tag on `len`. The precondition is supposed to be promoted from `!UnknownTagCheck([[i < len]], SecretTaint)`, but the promotion is currently not supported because the condition contains a local variable `i`.

Reply:
Meanwhile, put the precondition on the entire array and rely on TagPropagation::SubComponent to propagate the tag to the individual elements. (That too, might now work right now, so make an issue if need be.)

Reply:
When I tried to attach a tag to the array, I found it impossible to tag values with unsized types. Changed the parameters in the example to fixed-length arrays and submitted a PR #588 for the issue.